### PR TITLE
matchArgs: Add Not{Equal, Prefix, Postfix} in fd, file, path types

### PR DESCRIPTION
**Pull request by tpapagian**
FIXES: https://github.com/cilium/tetragon/issues/1310

### CODE REVIEW

**Review by jrfastab — APPROVED** [review]

File: bpf/process/types/basic.h [hunk 1]
```diff
@@ -697,26 +695,41 @@ struct file_buf {
...
 	char *values = (char *)filter + sizeof(struct file_buf); // all values are after the header
-	__u32 remain = filter->total_sz - sizeof(__u32) - sizeof(__u32);
-	int i, next;
+	__u32 next, remain = filter->total_sz - sizeof(__u32) - sizeof(__u32);
+	long i, match;
+
+	/* There are cases where file pointer may not contain a path.
+	 * An example is using an unnamed pipe. This is not a match.
+	 */
+	if (args->len == 0)
```
Comment (kkourt) [inline bpf/process/types/basic.h:707]: Haven't followed the code closely, but shouldn't we return  for the `not` operators? Maybe `goto: done_filter_file`?

File: bpf/process/types/basic.h [hunk 2]
```diff
@@ -697,26 +695,41 @@ struct file_buf {
...
-		if (!next)
+		struct string_buf *value = (struct string_buf *)values;
+
+		match = __filter_file_buf(value, args, filter->op);
+		if (filter->op == op_filter_neq || filter->op == op_filter_str_notprefix || filter->op == op_filter_str_notpostfix) {
+			if (match != FILTER_FILE_NOMATCH)
+				return 0;
+		} else if (match == FILTER_FILE_MATCH) {
 			return 1;
+		}
```
Comment (kkourt) [inline bpf/process/types/basic.h:722]: I'd personally write this as:

```
....
// the file matched, which means that we can now make a decision.
if (match == FILTER_FILE_MATCH) {
     return is_not_operator(filter->op) ? 0 : 1; 
}
```

Where:
```
bool is_not_operator(op) {
    return (filter->op == op_filter_neq || filter->op == op_filter_str_notprefix || filter->op == op_filter_str_notpostfix);
}
```

But that's just me :)

**Review by kkourt — COMMENTED** [review]
Thanks! I have some comments.


> The behaviour changes a bit when having multiple values. For operators Equal, Prefix, and Postfix the
> previous example will match when:
> [(arg OP val1) OR (arg OP val2) OR (arg OP val3)]
> 
> For operators NotEqual, NotPrefix, and NotPostfix the previous example will match when:
> [(arg OP val1) AND (arg OP val2) AND (arg OP val3)]

I think the right way to think of Not* operators is:

Equal: (arg == val1) || (arg == val2) || (arg == val3)
NotEqual: NOT(Equal)
 => NOT( (arg == val1) || (arg == val2) || (arg == val3) )
 => (arg != val1) && (arg != val2) && (arg != val3)
 
 So the behavior doesn't really change.

File: bpf/process/types/basic.h (hunk 1 above)
Comment (tpapagian) [reply bpf/process/types/basic.h:707]: In that case, we get an event where the path is an empty string. I don't find much value in having events like this when trying to monitor access to files. This is the reason I kept the same behavior for not operators. 

If this does not make sense, I can change that.

File: bpf/process/types/basic.h (hunk 2 above)
Comment (tpapagian) [reply bpf/process/types/basic.h]: Simplifies the code a bit. Done!

File: bpf/process/types/basic.h (hunk 1 above)
Comment (kkourt) [reply bpf/process/types/basic.h:712]: OK, that makes sense to me. Thanks!

**Review by kkourt — APPROVED** [review]

**Comment by tixxdz** [comment]
This broken main, how did it pass tests?

**Comment by tixxdz** [comment]
ah the kprobe_test doesn't have the observer package since it was moved so the tests need to rebase and use the new observerhelper package

**Comment by tpapagian** [comment]
> ah the kprobe_test doesn't have the observer package since it was moved so the tests need to rebase and use the new observerhelper package

Fix: https://github.com/cilium/tetragon/pull/1347

**Pull request merged**